proposals/pr-contrib-guide.md

```diff
@@ -0,0 +1,434 @@
+
+The maintainers will review the pull request and provide any comments, feedback,
+or requested changes. If changes are requested, the contributor should make the
+changes within [TBD] business days. If a pull request becomes stale because the
+contributor is no longer actively engaged or working on it for more than [TBD]
+days, it should be reviewed to see if it is suitable for the maintainers to
+finish the work or marked as stale and documented for another user to continue
+working on it.
+
+Once a pull request is complete with to the satisfaction of the maintainers, it
+is approved and can be merged and acknowledged as contribution from the
+community.
+
+The maintainers will prioritize completing PRs that have had a review over new
+review requests. If a pull request becomes stale but there is sufficient work to
+complete it, the maintainers will complete the pull request. If after review and
+the maintainers agree that the implementation is not as discussed or there is
+not significant demand or bandwidth to complete it, the issue may go back to the
+intake phase and be available for another user to contribute a change.
+
+If an issue has not been explicitly marked as being available for contribution,
+it should not be expected that a pull request will get a review.
+
+## Measuring success and making adjustments
+
+We acknowledge that the proposed timelines and queue sizes may need to be
+adjusted. We will quantify the process with metrics to determine if it is
+sustainable or if it can be improved. Specifically, we should measure:
+
+1. The number of issues in each state of the contribution process: available for
+   contribution, actively being contributed, ready for review.
+1. The number of popular and highly requested features.
+1. The time elapsed between each state of the contribution process: the start of
+   a contribution and when it is marked as ready for review, the time spent
+   waiting for review, and the time to accept or reject a contribution.
+1. The rate of acceptance and attrition for contributions.
+
+In addition, we will monitor qualitative feedback through comments left on
+issues and pull requests. We will review the feedback for common patterns to
+determine if an improvement to the documentation will alleviate contributor
+frustration. It may also indicate that we need to make changes to the process as
+well.
+
+## Managing the existing backlog
+
+As of 2021-11-02, there are 189 pull requests and 465 issues in the AWS CLI
+GitHub repository. 52 pull requests and 81 issues have not been updated in more
+than a year. 151 issues have upvotes or reactions from five or more users. [TBD]
+of the open pull requests are extremely outdated and would require rebasing,
+which complicates the process of accepting them. [TBD] of the open pull requests
+are directly related to an existing feature request.
+
+Based on these numbers, we propose to close all issues that:
```

Comment:
We should just be basing this off of the criteria for reaching ready for contribution state and just reference that definition here and make that the starting basis for this new process.

Specifically, we should be doing the following:

* Go through each PR. See if there is a corresponding issue that meets the criteria for "ready for contribution". If there is not, close the PR and tell them what needs to happen to reach  "ready for contribution". Once the issue reaches "ready for contribution" they can reopen the PR or create a new one.

* For PR's that have issues that are deemed "ready for contribution", evaluate whether they are "ready for review". If the PR is not ready to review, comment on what is left (i.e. rebase, get test passing) to reach "ready for review" state and give them X days to update the PR to acknowledge that they would still like to work on it. If they do not respond or say they no longer want to work on it, close PR and note in tracking GitHub issue that the issue is no longer being worked on and is still available for contribution.

* Sort the PRs "ready for review" and have the maintainers start working through the review queue based on this process.

But I think the key part to this is that our approach to addressing the currently open pull requests should match what the process is going forward and all wording in this section should be using the same terminology as in the section that described the new process (i.e. "ready for contribution", "ready for review"). Otherwise if we have a separate process, it is going to be difficult to manage the process for all of the preexisting PRs while trying to apply the new process to new PRs

Reply:
I'm not convinced that this is the best experience for users who opened PRs before this proposal. I think the same buffer for getting something "ready for review" should exist to get something "ready for contribution" to allow for an onboarding process for existing contributors. If we will aggressively close PRs, then we should commit to aggressively getting existing ones (based on their popularity) back to "ready for contribution" state.

The rationale for this is twofold:

1. We've previously told users to close an issue and just open a PR.
2. There are definitely some PRs with at least some components of the "ready for contribution" checklist completed.

Reply:
I do agree with matching the terminology!

Reply:
Think this is resolved per offline conversation, resolving.